File: a3_finance/a3_finance/report/professional_tax_summary/professional_tax_summary.js
// Copyright (c) 2025, Acube and contributors
// For license information, please see license.txt

frappe.query_reports["Professional Tax Summary"] = {

	onload: function (report) {

		// ---- PRINT BUTTON ----
		report.page.add_inner_button("Print Report", function () {
			print_professional_tax_report(report);
		});

		// Hide Sl No column visually
		setTimeout(() => {
			const wrapper = report.page.wrapper;
			wrapper.find(".dt-header .dt-cell--col-0").hide();
			wrapper.find(".dt-body .dt-cell--col-0").hide();
		}, 300);
	},

	after_datatable_render: function (datatable) {
		datatable.columnmanager.columns[0].hidden = true;
		datatable.refresh();
	},

	filters: [
		{
			fieldname: "company",
			label: "Company",
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company")
		},
		{
			fieldname: "month",
			label: "Month",
			fieldtype: "Select",
			options: ["01","02","03","04","05","06","07","08","09","10","11","12"],
			default: frappe.datetime.get_today().split("-")[1]
		},
		{
			fieldname: "year",
			label: "Year",
			fieldtype: "Int",
			default: frappe.datetime.get_today().split("-")[0]
		},
		{
			fieldname: "prepared_by",
			label: "Prepared By",
			fieldtype: "Link",
			options: "Employee"
		}
	],

	// ---------------- PRINTABLE HTML ----------------
	printable_html: async function (report) {

		let filters = frappe.query_report.get_filter_values();
		let data = report.data || [];
		let columns = report.columns || [];

		const month_names = [
			"January","February","March","April","May","June",
			"July","August","September","October","November","December"
		];

		let month_name = month_names[parseInt(filters.month) - 1];
		let company_name = filters.company || "All Companies";
		let prepared_by = filters.prepared_by || "";
		let prepared_by_name = prepared_by;
		let prepared_by_designation = "";

		if (prepared_by) {
			const emp_details = await frappe.db.get_value(
				"Employee",
				prepared_by,
				["employee_name", "designation"]
			);
			prepared_by_name = emp_details?.message?.employee_name || prepared_by;
			prepared_by_designation = emp_details?.message?.designation || "";
		}

		let html = `
<style>
@page {
	margin: 10mm;
}

@media print {
	body {
		margin: 0;
	}

	table {
		border-collapse: collapse;
		width: 99%;
		font-size: 11px;
	}

	th, td {
		border: 1px solid #000;
		padding: 4px;
		box-sizing: border-box;
	}

	th {
		text-align: center;
		font-weight: bold;
	}

	td {
		text-align: left;
	}

	td.num {
		text-align: right;
	}

	tr.total-row {
		font-weight: bold;
	}
}
</style>

<div style="text-align:center; margin-bottom:15px;">
<div style="font-size:18px; font-weight:bold;">
    ${company_name}
</div>

<div style="font-size:13px;">CHACKAI, BEACH.P.O., AIRPORT ROAD, THIRUVANANTHAPURAM</div>
	<div style="font-size:18px; font-weight:bold;">
		Professional Tax Summary
	</div>
	<div style="font-size:14px; margin-top:4px;">
		
		Month: <b>${month_name}</b> &nbsp; | &nbsp;
		Year: <b>${filters.year}</b>
	</div>
</div>

<table>
	<thead>
		<tr>
`;

		columns.forEach(col => {
			html += `<th>${col.label}</th>`;
		});

		html += `</tr></thead><tbody>`;

		data.forEach(row => {

			let is_total = row.ec === "Total";
			html += `<tr class="${is_total ? 'total-row' : ''}">`;

			columns.forEach(col => {

				// Hide Sl No value for Total row
				if (col.fieldname === "sl_no" && is_total) {
					html += `<td></td>`;
					return;
				}

				let value = row[col.fieldname] ?? "";
				let cls = col.fieldtype === "Currency" ? "num" : "";

				html += `<td class="${cls}">
					${frappe.format(value, col)}
				</td>`;
			});

			html += `</tr>`;
		});

		html += `
	</tbody>
</table>

<div style="margin-top:35px; width:280px;">
	<div style="font-size:11px; font-weight:bold; margin-bottom:6px;">Prepared By</div>
	<div style="height:50px;"></div>
	
		<div><b>${prepared_by_name}</b></div>
		<div>${prepared_by_designation}</div>
	</div>
</div>




`;

		return html;
	}
};


// ---------------- CUSTOM PRINT FUNCTION ----------------
async function print_professional_tax_report(report) {

	let html = await frappe.query_reports["Professional Tax Summary"]
		.printable_html(report);

	let print_window = window.open("", "PRINT", "height=700,width=900");

	print_window.document.write(html);
	print_window.document.close();
	print_window.focus();
	print_window.print();
	print_window.close();
}

File: a3_finance/a3_finance/report/professional_tax_summary/professional_tax_summary.py
# Copyright (c) 2025, Acube and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import get_first_day, get_last_day


def execute(filters=None):
	columns = get_columns()
	data = get_data(filters)
	return columns, data

def get_columns():
	return [
		{"label":"Sl No", "fieldname": "sl_no", "fieldtype": "Int", "width": 50},
		{"label":"EC", "fieldname": "ec", "fieldtype": "Data", "width": 120},
		{"label":"Employee Name", "fieldname": "employee_name", "fieldtype" : "Data", "width": 200},
		{"label":"Basic Pay 4 Months" , "fieldname": "basic_pay", "fieldtype": "Currency", "width": 120},
		{"label": "Normal Basic Pay", "fieldname": "normal_basic_pay", "fieldtype": "Currency", "width": 150},
		{"label": "Estimated for 2 Months", "fieldname": "estimated_basic_pay", "fieldtype": "Currency", "width": 180},
		{"label":"BP 6 Months", "fieldname": "bp_6_months", "fieldtype": "Currency", "width": 120},
		{"label":"VDA 4 Months", "fieldname": "vda", "fieldtype": "Currency", "width": 120},
		{"label": "Normal VDA", "fieldname": "normal_vda", "fieldtype": "Currency", "width": 120},
		{"label": "Estimated for 2 Months", "fieldname": "estimated_vda", "fieldtype": "Currency", "width": 180},
		{"label":"VDA 6 Months", "fieldname": "vda_6_months", "fieldtype": "Currency", "width": 120},
		{"label":"Total (BP + VDA) 6 Months", "fieldname": "total_bp_vda", "fieldtype": "Currency", "width": 180},
		{"label":"Professional Tax", "fieldname": "professional_tax", "fieldtype": "Currency", "width": 120},
	]

def get_data(filters=None):
	month = int(filters.get("month"))
	year = int(filters.get("year"))
	company = filters.get("company")
	start_date = get_first_day(f"{year}-{month}-01")

	data = []
	sl_no = 1
	total_bp=0
	total_normal_bp=0
	total_estimated_bp=0
	total_bp_6_months=0
	total_vda=0
	total_normal_vda=0
	total_estimated_vda=0
	total_vda_6_months=0
	total_total_bp_vda=0
	total_professional_tax=0
	salary_slip = frappe.db.sql(
		"""
		SELECT name, employee, employee_name
		FROM `tabSalary Slip`
		WHERE start_date = %s
		  AND docstatus = 1
		  AND (%s IS NULL OR company = %s)""",
		(start_date, company, company),
		as_dict=True,
	)
	for slip in salary_slip:
		proffessional_tax = frappe.db.sql(""" SELECT amount FROM `tabSalary Detail` WHERE parent = %s AND parentfield="deductions" AND salary_component = "Professional Tax" """, slip.name	)
		if proffessional_tax:
			bp= frappe.db.sql(""" SELECT amount,year_to_date,custom_actual_amount FROM `tabSalary Detail` WHERE parent = %s AND parentfield="earnings" AND salary_component = 'Basic Pay' """, slip.name	)
			vda = frappe.db.sql(""" SELECT amount,year_to_date,custom_actual_amount FROM `tabSalary Detail` WHERE parent = %s AND parentfield="earnings" AND salary_component = 'Variable DA' """, slip.name	)
			data.append({
			"sl_no": sl_no,
			"ec": slip.employee,
			"employee_name": slip.employee_name,
			"basic_pay": bp[0][1] if bp else 0,
			"normal_basic_pay": bp[0][2] if bp else 0,
			"estimated_basic_pay": (bp[0][2]*2) if bp else 0,
			"bp_6_months": (bp[0][1] + (bp[0][2])*2) if bp else 0,
			"vda": vda[0][1] if vda else 0,
			"normal_vda": vda[0][2] if vda else 0,
			"estimated_vda": (vda[0][2])*2 if	 vda else 0,
			"vda_6_months": (vda[0][1] + (vda[0][2])*2) if vda else 0,
			"total_bp_vda": ((bp[0][1] + (bp[0][2])*2) + (vda[0][1] + (vda[0][2])*2)) if bp and vda else 0,
			"professional_tax": proffessional_tax[0][0] if proffessional_tax else 0
			})
			total_bp += bp[0][1] if bp else 0
			total_normal_bp += bp[0][2] if bp else 0
			total_estimated_bp += (bp[0][2]*2) if bp else 0
			total_bp_6_months += (bp[0][1] + (bp[0][2])*2) if bp else 0
			total_vda += vda[0][1] if vda else 0
			total_normal_vda += vda[0][2] if vda else 0
			total_estimated_vda += (vda[0][2])*2 if	 vda else 0
			total_vda_6_months += (vda[0][1] + (vda[0][2])*2) if vda else 0
			total_total_bp_vda += ((bp[0][1] + (bp[0][2])*2) + (vda[0][1] + (vda[0][2])*2)) if bp and vda else 0
			total_professional_tax += proffessional_tax[0][0] if proffessional_tax else 0
			sl_no += 1
	data.append({
		"sl_no": "",
		"ec": "Total",
		"employee_name": "",
		"basic_pay": total_bp,
		"normal_basic_pay": total_normal_bp,
		"estimated_basic_pay": total_estimated_bp,
		"bp_6_months": total_bp_6_months,
		"vda": total_vda,
		"normal_vda": total_normal_vda,
		"estimated_vda": total_estimated_vda,
		"vda_6_months": total_vda_6_months,
		"total_bp_vda": total_total_bp_vda,
		"professional_tax": total_professional_tax
	})
	return data
